### PR TITLE
[cmake] Setup C++ Standard flags from CMAKE_CXX_STANDARD and add C++ standard to root-config/RootUseFile.cmake

### PR DESCRIPTION
**Pull request by oshadura**
Fixes: https://github.com/root-project/root/issues/6384

I see during "random" TU compilation flags are setup properly:
```[ 78%] Building CXX object roofit/roofitcore/CMakeFiles/RooFitCore.dir/src/RooAbsMCStudyModule.cxx.o
cd /home/oksana/CERN_sources/root/builds-no-cxx14/roofit/roofitcore && /usr/bin/c++  -I/home/oksana/CERN_sources/root/roofit/roofitcore/inc -I/home/oksana/CERN_sources/root/core/unix/inc -I/home/oksana/CERN_sources/root/core/foundation/v7/inc -I/home/oksana/CERN_sources/root/core/base/v7/inc -I/home/oksana/CERN_sources/root/core/clingutils/inc -I/home/oksana/CERN_sources/root/core/textinput/inc -I/home/oksana/CERN_sources/root/core/thread/inc -I/home/oksana/CERN_sources/root/core/zip/inc -I/home/oksana/CERN_sources/root/core/rint/inc -I/home/oksana/CERN_sources/root/core/clib/inc -I/home/oksana/CERN_sources/root/core/meta/inc -I/home/oksana/CERN_sources/root/core/gui/inc -I/home/oksana/CERN_sources/root/core/cont/inc -I/home/oksana/CERN_sources/root/core/foundation/inc -I/home/oksana/CERN_sources/root/core/base/inc -I/home/oksana/CERN_sources/root/builds-no-cxx14/ginclude -I/home/oksana/CERN_sources/root/hist/hist/inc -I/home/oksana/CERN_sources/root/math/mathcore/v7/inc -I/home/oksana/CERN_sources/root/math/mathcore/inc -I/home/oksana/CERN_sources/root/core/imt/inc -I/home/oksana/CERN_sources/root/math/matrix/inc -I/home/oksana/CERN_sources/root/graf2d/graf/inc -I/home/oksana/CERN_sources/root/io/io/v7/inc -I/home/oksana/CERN_sources/root/io/io/inc -I/home/oksana/CERN_sources/root/io/io/res -I/home/oksana/CERN_sources/root/core/clib/res -I/usr/include/freetype2 -I/home/oksana/CERN_sources/root/tree/tree/inc -I/home/oksana/CERN_sources/root/math/minuit/inc -I/home/oksana/CERN_sources/root/math/foam/inc -I/home/oksana/CERN_sources/root/math/smatrix/inc -I/home/oksana/CERN_sources/root/net/net/inc -Wno-implicit-fallthrough -Wno-noexcept-type -pipe  -Wshadow -Wall -W -Woverloaded-virtual -fsigned-char -pthread -O3 -DNDEBUG -fPIC -std=c++14 -o CMakeFiles/RooFitCore.dir/src/RooAbsMCStudyModule.cxx.o -c /home/oksana/CERN_sources/root/roofit/roofitcore/src/RooAbsMCStudyModule.cxx
```
For roottest/rootbench and etc., could be useful next output:
```
~ root-config --cxxstandard
14
 ```

### CODE REVIEW

File: config/root-config.in [hunk 1]
```diff
@@ -828,6 +831,7 @@ while test $# -gt 0; do
       echo "  --python-version      Print the Python version used by ROOT"
       echo "  --python2-version     Print the Python2 version used by PyROOT"
       echo "  --python3-version     Print the Python3 version used by PyROOT"
+      echo "  --cxxstandard         Print the C++ standard used to compile ROOT"
```
Comment (Axel-Naumann) [inline config/root-config.in:834]: I don't understand the order of the flags here. Nor the order of the flags in the `--help` output. Nor their relation.

But that's not this PR's business... I'd certainly appreciate e.g. an alphabetical order for both in a subsequent PR :-)

**Review by Axel-Naumann — APPROVED** [review]

**Comment by oshadura** [comment]
@phsft-bot build with flags -DCTEST_TEST_EXCLUDE_NONE=On

**Comment by oshadura** [comment]
@Axel-Naumann I don't understand why locally everything works, but not at Jenkins....Could it be connected with that some executables/cache are not cleaned properly between PRs?

**Comment by oshadura** [comment]
@pcanal actually with my PR, looks like we don't have C++ standard flags in `compiledata.h`. Is it used for Aclic?
```
/* This file is automatically generated */
#define BUILD_ARCH "linuxx8664gcc"
#define BUILD_NODE "Linux oksana-xps139380 5.7.17-2-MANJARO #1 SMP PREEMPT Sat Aug 22 14:58:17 UTC 2020 x86_64 GNU/Linux"
#define CXX "c++"
#define COMPILER "/usr/bin/c++"
#define COMPILERVERS "gcc1020"
#define COMPILERVERSSTR "c++ (GCC) 10.2.0"
#define MAKESHAREDLIB  "cd $BuildDir ; c++ -fPIC -c $Opt -Wno-implicit-fallthrough -Wno-noexcept-type -pipe -W -Woverloaded-virtual -fsigned-char -pthread $IncludePath $SourceFiles ; c++ $Opt $ObjectFiles -shared  $LinkedLibs -o $SharedLib"
#define MAKEEXE "cd $BuildDir ; c++ -c $Opt -Wno-implicit-fallthrough -Wno-noexcept-type -pipe -W -Woverloaded-virtual -fsigned-char -pthread $IncludePath $SourceFiles; c++ $Opt $ObjectFiles  -o $ExeName $LinkedLibs"
#define CXXOPT "-O3 -DNDEBUG"
#define CXXDEBUG "-g"
#define ROOTBUILD ""
#define LINKEDLIBS "-L$ROOTSYS/lib -lCore -lRint "
#define INCLUDEPATH "-I$ROOTSYS/include"
#define OBJEXT "o"
#define SOEXT "so"
```

**Comment by pcanal** [comment]
> compiledata.h. Is it used for Aclic?

Yes, this is ACLiC's only source of information for the flags.

**Comment by oshadura** [comment]
@pcanal Thanks, this explains everything (almost):)

**Comment by guitargeek** [comment]
Rebased on master to fix conflicts. Let's see what the status of this PR is.

**Review by guitargeek — APPROVED** [review]
Thanks @oshadura for your work! It was almost ready to go, only one remaining problem that I needed to fix up:

In the `RootUseFile.cmake` in, I replaced this line:
```cmake
set(CMAKE_CXX_STANDARD ${CXX_STANDARD_STRING})
```
with the correct way to propagate this information (via the `ROOT_COMPILER_FLAG_HINTS`, just like for the other `CXX_FLAGS` etc.).

File: CMakeLists.txt [hunk 2]
```diff
@@ -759,18 +759,19 @@ endif()
 cmake_host_system_information(RESULT PROCESSOR QUERY PROCESSOR_DESCRIPTION)
 
 message(STATUS "ROOT Configuration \n
-System          ${CMAKE_SYSTEM}
-Processor       ${PROCESSOR} (${CMAKE_SYSTEM_PROCESSOR})
-Build type      ${CMAKE_BUILD_TYPE}
-Install path    ${CMAKE_INSTALL_PREFIX}
-Compiler        ${CMAKE_CXX_COMPILER_ID} ${CMAKE_CXX_COMPILER_VERSION}
+System:          ${CMAKE_SYSTEM}
```
Comment (Axel-Naumann) [inline CMakeLists.txt:762]: This extra indentation doesn't seem to be needed?

File: cmake/modules/CheckCompiler.cmake [hunk 3]
```diff
@@ -249,11 +246,3 @@ check_cxx_source_compiles(
 #endif
 int main() {}
 " GLIBCXX_USE_CXX11_ABI)
-
-#---Print the final compiler flags--------------------------------------------------------------------
```
Comment (Axel-Naumann) [inline cmake/modules/CheckCompiler.cmake:253]: Why is this removed?

File: cmake/modules/RootConfiguration.cmake [hunk 4]
```diff
@@ -718,9 +721,11 @@ if(WIN32)
   # We cannot use the compiledata.sh script for windows
   configure_file(${CMAKE_SOURCE_DIR}/cmake/scripts/compiledata.win32.in ${CMAKE_BINARY_DIR}/ginclude/compiledata.h NEWLINE_STYLE UNIX)
 else()
+  # Needed by ACLIC, while in ROOT we are using everywhere C++ standard via CMake features that are requested to build target
+  set(CMAKE_CXX_ACLIC_FLAGS "${CMAKE_CXX_FLAGS} ${CMAKE_CXX${CMAKE_CXX_STANDARD}_STANDARD_COMPILE_OPTION}")
```
Comment (Axel-Naumann) [inline cmake/modules/RootConfiguration.cmake:725]: Event without this, i.e. in current master, my `ginclude/comiledata.h` contains `-std=c++17`. So this doesn't seem needed?

File: cmake/modules/RootMacros.cmake [hunk 5]
```diff
@@ -915,9 +915,7 @@ function(ROOT_LINKER_LIBRARY library)
     target_link_libraries(${library} PUBLIC ${ARG_LIBRARIES} ${ARG_DEPENDENCIES})
   endif()
 
-  if(DEFINED CMAKE_CXX_STANDARD)
-    target_compile_features(${library} INTERFACE cxx_std_${CMAKE_CXX_STANDARD})
-  endif()
+  set_target_properties(${library} PROPERTIES CXX_STANDARD ${CMAKE_CXX_STANDARD})
```
Comment (Axel-Naumann) [inline cmake/modules/RootMacros.cmake:918]: That change should be announced in the release notes. I think it's fine to remove the old one; I don't see why we would want to add the new call to `set_target_properties`, because according to [the doc](https://cmake.org/cmake/help/latest/prop_tgt/CXX_STANDARD.html)
> This property is initialized by the value of the [CMAKE_CXX_STANDARD](https://cmake.org/cmake/help/latest/variable/CMAKE_CXX_STANDARD.html#variable:CMAKE_CXX_STANDARD) variable if it is set when a target is created.

File: cmake/modules/RootMacros.cmake [hunk 6]
```diff
@@ -1092,7 +1090,7 @@ function(ROOT_OBJECT_LIBRARY library)
   # target. Doing so is actively harmful for the modules build because it
   # creates extra module variants, and not useful because we don't use these
   # macros.
-  set_target_properties(${library} PROPERTIES DEFINE_SYMBOL "")
+  set_target_properties(${library} PROPERTIES DEFINE_SYMBOL "" CXX_STANDARD ${CMAKE_CXX_STANDARD})
```
Comment (Axel-Naumann) [inline cmake/modules/RootMacros.cmake:1093]: Likely not needed, see above

File: cmake/modules/RootMacros.cmake [hunk 7]
```diff
@@ -1141,7 +1139,7 @@ function(ROOT_MODULE_LIBRARY library)
   ROOT_GET_SOURCES(lib_srcs src ${ARG_UNPARSED_ARGUMENTS})
   add_library(${library} SHARED ${lib_srcs})
   add_dependencies(${library} move_headers)
-  set_target_properties(${library}  PROPERTIES ${ROOT_LIBRARY_PROPERTIES})
```
Comment (Axel-Naumann) [inline cmake/modules/RootMacros.cmake:1144]: Likely not needed, see above

File: cmake/modules/RootMacros.cmake [hunk 8]
```diff
@@ -1438,7 +1436,7 @@ function(ROOT_EXECUTABLE executable)
     set_target_properties(${executable} PROPERTIES SUFFIX "")
   endif()
   set_property(GLOBAL APPEND PROPERTY ROOT_EXPORTED_TARGETS ${executable})
-  set_target_properties(${executable} PROPERTIES OUTPUT_NAME ${executable_name})
+  set_target_properties(${executable} PROPERTIES OUTPUT_NAME ${executable_name} CXX_STANDARD ${CMAKE_CXX_STANDARD})
```
Comment (Axel-Naumann) [inline cmake/modules/RootMacros.cmake:1439]: Likely not needed, see above

File: cmake/modules/RootMacros.cmake [hunk 9]
```diff
@@ -1845,6 +1843,8 @@ function(ROOT_ADD_GTEST test_suite)
     target_include_directories(${test_suite} PRIVATE ${ARG_INCLUDE_DIRS})
   endif(ARG_INCLUDE_DIRS)
 
+  set_property(TARGET ${test_suite} PROPERTY CXX_STANDARD ${CMAKE_CXX_STANDARD})
```
Comment (Axel-Naumann) [inline cmake/modules/RootMacros.cmake:1846]: Likely not needed, see above

**Review by Axel-Naumann — CHANGES_REQUESTED** [review]
We might actually only need a very small subset of these changes. I'd prefer to wait for @bellenot to be back and to have a look.

File: cmake/modules/CheckCompiler.cmake (hunk 3 above)
Comment (guitargeek) [reply cmake/modules/CheckCompiler.cmake:253]: It's partially redundant with what is already printed in the top-level `CMakeLists.txt.

However, it was not *completely* redundant. I updated the other message print to include all the information printed here.

File: cmake/modules/RootConfiguration.cmake (hunk 4 above)
Comment (guitargeek) [reply cmake/modules/RootConfiguration.cmake]: Yes, seems like it! I reverted this change.

File: cmake/modules/RootMacros.cmake (hunk 5 above)
Comment (guitargeek) [reply cmake/modules/RootMacros.cmake]: Ok, let's not add the new line. But why would you announce this change in the release notes? The behavior is not changed by removing this, because the C++ standard is taken form `CMAKE_CXX_STANDARD` also by default?

**Comment by guitargeek** [comment]
Hi Axel, thanks for your review! I have followed up on it, because Oksana is not working on ROOT anymore and I think this is still a relevant improvement.

You have any more comments?

File: cmake/modules/RootConfiguration.cmake (hunk 4 above)
Comment (guitargeek) [reply cmake/modules/RootConfiguration.cmake:723]: Actually it was necessary. Without this PR, you are right, it contains it because of this hack in `CheckCompiler.cmake`:
```cmake
set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${CMAKE_CXX${CMAKE_CXX_STANDARD}_STANDARD_COMPILE_OPTION}")
```

But this PR is removing that hack, so now the standard is not in the `CMAKE_CXX_FLAGS` anymore and we need to add it back for `compiledata.sh`.

**Comment by guitargeek** [comment]
Rebased onto master to trigger the CI.

**Review by bellenot — APPROVED** [review]
LGTM. `root-config.bat.in` changes are missing, but I'll add them later
